File: darwin/video_watermark/Sources/video_watermark/VideoWatermarkPlugin.swift
#if os(macOS)
import Cocoa
import FlutterMacOS
#else
import Flutter
import UIKit
#endif

public class VideoWatermarkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(macOS)
    let messenger = registrar.messenger
    #else
    let messenger = registrar.messenger()
    #endif
    let channel = FlutterMethodChannel(name: "video_watermark", binaryMessenger: messenger)
    let instance = VideoWatermarkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      #if os(macOS)
      result("macOS " + ProcessInfo.processInfo.operatingSystemVersionString)
      #else
      result("iOS " + UIDevice.current.systemVersion)
      #endif
    case "addWatermark":
      let args = call.arguments as? [String: Any]
      let videoPath = args?["videoPath"] as? String ?? "Unknown"
      let watermark = args?["watermark"] as? [String: Any]
      let position = args?["position"] as? [String: Any]

      let watermarkStr = watermark != nil ? String(describing: watermark!) : "unknown watermark"
      let posStr = position != nil ? String(describing: position!) : "default"

      var extras = ""
      if let wmStart = args?["watermarkStartTime"] as? Int { extras += "wm_start: \(wmStart) " }
      if let wmEnd = args?["watermarkEndTime"] as? Int { extras += "wm_end: \(wmEnd) " }
      if let rw = args?["resizeWidth"] as? Int { extras += "resizeW: \(rw) " }
      if let rh = args?["resizeHeight"] as? Int { extras += "resizeH: \(rh) " }
      if let trimStart = args?["trimStart"] as? Int { extras += "trim_start: \(trimStart) " }
      if let trimEnd = args?["trimEnd"] as? Int { extras += "trim_end: \(trimEnd) " }
      if let quality = args?["compressionQuality"] as? Double { extras += "quality: \(quality)" }

      #if os(macOS)
      result("macOS watermarked " + videoPath + " with " + watermarkStr + " at " + posStr + " (" + extras + ")")
      #else
      result("iOS watermarked " + videoPath + " with " + watermarkStr + " at " + posStr + " (" + extras + ")")
      #endif
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/widle/video_watermark/VideoWatermarkPlugin.kt
package com.widle.video_watermark

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class VideoWatermarkPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "video_watermark")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
        } else if (call.method == "addWatermark") {
            val videoPath = call.argument<String>("videoPath")
            val watermark = call.argument<Map<String, Any>>("watermark")
            val position = call.argument<Map<String, Any>>("position")
            val watermarkStartTime = call.argument<Int>("watermarkStartTime")
            val watermarkEndTime = call.argument<Int>("watermarkEndTime")
            val resizeWidth = call.argument<Int>("resizeWidth")
            val resizeHeight = call.argument<Int>("resizeHeight")
            val trimStart = call.argument<Int>("trimStart")
            val trimEnd = call.argument<Int>("trimEnd")
            val compressionQuality = call.argument<Double>("compressionQuality")

            val watermarkStr = if (watermark != null) watermark.toString() else "unknown watermark"
            val posStr = if (position != null) position.toString() else "default"

            var extras = ""
            if (watermarkStartTime != null) extras += "wm_start: $watermarkStartTime "
            if (watermarkEndTime != null) extras += "wm_end: $watermarkEndTime "
            if (resizeWidth != null) extras += "resizeW: $resizeWidth "
            if (resizeHeight != null) extras += "resizeH: $resizeHeight "
            if (trimStart != null) extras += "trim_start: $trimStart "
            if (trimEnd != null) extras += "trim_end: $trimEnd "
            if (compressionQuality != null) extras += "quality: $compressionQuality"

            result.success("Android watermarked $videoPath with $watermarkStr at $posStr ($extras)")
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
